### PR TITLE
Give error when we find a circular reference

### PR DESCRIPTION
**Pull request by pksunkara**
This PR gives a parser error when the parser finds a circular reference among the named types. The old #315 only takes care of circular *inheritance*.

Note: It was a bit tricky to think about how to do this which is why it took me a bit of time.

@zdne @klokane Please review and merge.

### CODE REVIEW

File: src/SectionParserData.h [hunk 1]
```diff
@@ -51,6 +51,12 @@ namespace snowcrash {
         /** Table mapping named type to sub types */
         mson::NamedTypeInheritanceTable namedTypeInheritanceTable;
 
+        /** Table mapping named types to their dependent named types */
+        mson::NamedTypeDependencyTable namedTypeDependencyTable;
+
+        /** Variable to store the current named type */
+        mson::Literal namedTypeContext;
```
Comment (pksunkara) [inline src/SectionParserData.h]: I didn't like writing this here at all, but it was necessary. Maybe there is a better way.

File: src/BlueprintParser.h [hunk 2]
```diff
@@ -357,14 +357,37 @@ namespace snowcrash {
                 return;
             }
 
-            // Otherwise, add the respective entries to the tables
-            if (typeDefinition.typeSpecification.name.base != mson::UndefinedTypeName) {
+            mson::BaseTypeName baseTypeName = typeDefinition.typeSpecification.name.base;
+
+            // Initialize an entry in the dependency table if not found
+            if (pd.namedTypeDependencyTable.find(identifier) != pd.namedTypeDependencyTable.end()) {
```
Comment (zdne) [inline src/BlueprintParser.h]: Better design: in a future this really needs to be split into multiple functions – this one is already 200 lines long and it is quite hard to follow what is happening here.

Comment (pksunkara) [inline src/BlueprintParser.h]: Yeah, I agree this needs to be split into multiple functions. This is only 100 lines, am I missing something else?

File: src/BlueprintParser.h [hunk 3]
```diff
@@ -383,43 +406,98 @@ namespace snowcrash {
...
             }
 
+            // Check for circular references
+            std::set<mson::Literal> deps = pd.namedTypeDependencyTable[subType];
+
+            if (deps.find(subType) != deps.end()) {
+
+                // ERR: A named type is circularly referenced
+                std::stringstream ss;
+                ss << "base type '" << subType << "' circularly referencing itself is not allowed";
```
Comment (zdne) [inline src/BlueprintParser.h]: Not sure where this is exactly being printed out. But I would say something like "recursive base type XXXX" , I would definitely skip the "is not allowed" part

File: src/BlueprintParser.h [hunk 4]
```diff
@@ -430,9 +508,9 @@ namespace snowcrash {
                 mson::NamedTypeInheritanceTable::iterator inhIt = pd.namedTypeInheritanceTable.find(superType);
 
                 // Check for recursive MSON definitions
-                if (inhIt == pd.namedTypeInheritanceTable.end() || inheritanceNodes.find(inhIt->second.first) != inheritanceNodes.end()) {
+                if (inhIt == pd.namedTypeInheritanceTable.end()) {
```
Comment (zdne) [inline src/BlueprintParser.h]: So, since this condition is loosened – can't this lead into additional problems?

File: src/MSONUtility.h [hunk 5]
```diff
@@ -121,26 +121,28 @@ namespace mson {
      *
      * \param subject String which represents the type name
      * \param typeName MSON Type Name
+     * \param notBaseType If true, will be parsed as a symbol
      */
     inline void parseTypeName(const std::string& subject,
-                              TypeName& typeName) {
+                              TypeName& typeName,
+                              bool notBaseType = false) {
```
Comment (zdne) [inline src/MSONUtility.h]: the name is not really fortunate – having `not` in the name and then `!` condition it is really hard to read...

File: src/BlueprintParser.h (hunk 4 above)
Comment (pksunkara) [inline src/BlueprintParser.h]: Nope. The earlier `Check for circular references` will cover it.

Comment (pksunkara) [inline src/BlueprintParser.h]: This PR is basically a refactor over the previous implementation.

File: src/MSONUtility.h [hunk 6]
```diff
@@ -418,6 +420,54 @@ namespace mson {
...
+
+        std::set<mson::Literal> dependencyDeps = pd.namedTypeDependencyTable[dependency];
+
+        // First, check if it is circular reference between them
+        if (dependent == dependency ||
+            dependencyDeps.find(dependent) != dependencyDeps.end()) {
+
+            // ERR: Dependency named type circular references itself
+            std::stringstream ss;
+            ss << "base type '" << dependent << "' circularly referencing itself is not allowed";
```
Comment (zdne) [inline src/MSONUtility.h]: again see my comment on this message above, also if they are the same error manifested by the same message a reuse of the message would be appropriate

File: src/SectionParserData.h (hunk 1 above)
Comment (zdne) [inline src/SectionParserData.h]: Hmm so this is storing the context – name of the named type being parsed – to store dependencies for later?

Comment (pksunkara) [inline src/SectionParserData.h]: Yup.

File: test/test-BlueprintParser.cc [hunk 7]
```diff
@@ -712,7 +727,146 @@ TEST_CASE("Report error when data Structure inheritance graph contains a cycle",
     SectionParserHelper<Blueprint, BlueprintParser>::parse(source, BlueprintSectionType, blueprint, ExportSourcemapOption, Models(), &blueprint);
 
     REQUIRE(blueprint.report.error.code == MSONError);
-    SourceMapHelper::check(blueprint.report.error.location, 28, 9);
+    SourceMapHelper::check(blueprint.report.error.location, 19, 9);
```
Comment (zdne) [inline test/test-BlueprintParser.cc]: Why did the location changed?

Comment (pksunkara) [inline test/test-BlueprintParser.cc]: Previously, parser used to resolve all the inheritance tables before returning error which is basically the last error occurring. Now, the parser returns error as soon as it encounters when resolving the inheritance tables. https://github.com/apiaryio/snowcrash/pull/317/files#diff-da3aa8ccabc921ae656c97332b64e901R425

File: src/BlueprintParser.h (hunk 2 above)
Comment (zdne) [inline src/BlueprintParser.h]: Oh lost in the diff. It is "just" something over 100 (but thats still too long) – OK to leave for now tho.

**Comment by pksunkara** [comment]
Updated.

File: src/MSONUtility.h [hunk 8]
```diff
@@ -121,26 +121,28 @@ namespace mson {
      *
      * \param subject String which represents the type name
      * \param typeName MSON Type Name
+     * \param isBaseType If false, will be parsed as a symbol
      */
     inline void parseTypeName(const std::string& subject,
-                              TypeName& typeName) {
+                              TypeName& typeName,
+                              bool isBaseType = true) {
```
Comment (klokane) [inline src/MSONUtility.h]: Better solution for readability is change condition:
```
if(!(isBaseType) {
  typeName.symbol = parseSymbol(subject)
} else {
  if (subject == "boolean") {
    typeName.base = BooleanTypeName;  
  } else if(...) {
  }   
}
```

And idealy move `else` into own method

**Comment by zdne** [comment]
cc @kylef
